### PR TITLE
Issue 1769 - part 1 - Adding of handling for AggregateExceptions

### PR DESCRIPTION
**Pull request by denis-tikhomirov**
Added handling of AggregateException exceptions generated in agent code.
Currently AggregateException exceptions are displayed alike "System.AggregateException: One or more errors occurred". The changes correct the layout to following:
`One or several exceptions have been occurred.`
`<text of exception 1>`
`<text of exception 2>`

---

Fix was tested manually by throwing of new AggregateException before call of handler implemented in the PR:
- place in code - https://github.com/microsoft/azure-pipelines-agent/blob/d487d723a7f44a6b95dc4b035736de99cdf59898/src/Agent.Worker/JobRunner.cs#L383-L384
- added lines:
   - `Exception exc1 = new Exception("InnerTestException01");`
   - `Exception exc2 = new Exception("InnerTestException02");`
   - `Exception[] excls = { exc1, exc2 };`
   - `throw new AggregateException("AggregateTestException01", excls );`

Result - info in Worker log:
`[2021-11-26 08:13:23Z ERR  JobRunner] One or several exceptions have been occurred.`
`[2021-11-26 08:13:23Z ERR  JobRunner] System.Exception: InnerTestException01`
`[2021-11-26 08:13:23Z ERR  JobRunner] System.Exception: InnerTestException02`

### CODE REVIEW

File: src/Agent.Listener/Agent.cs [hunk 1]
```diff
@@ -464,6 +465,17 @@ private async Task<int> RunAsync(AgentSettings settings, bool runOnce = false)
                                 Trace.Error($"Received message {message.MessageId} with unsupported message type {message.MessageType}.");
                             }
                         }
+                        catch (Exception e)
+                        {
+                            if (e is AggregateException)
```
Comment (anatolybolshakov) [inline src/Agent.Listener/Agent.cs:470]: Can we move condition to the common function too?

Comment (denis-tikhomirov) [reply src/Agent.Listener/Agent.cs:470]: No, statement ELSE can have different set of commands like "Trace.Debug", "execution.Context.Error", "throw", etc. So I think it's better to keep IF statement outside of handler.

File: src/Agent.Listener/Agent.cs [hunk 2]
```diff
@@ -464,6 +464,26 @@ private async Task<int> RunAsync(AgentSettings settings, bool runOnce = false)
...
+                            {
+                                Trace.Error("One or several exceptions have been occurred.");
+
+                                int i = 0;
+                                foreach (var ex in ((AggregateException)e).Flatten().InnerExceptions)
+                                {
+
+                                    i++;
+                                    Trace.Error($"InnerException #{i}");
+                                    Trace.Error(ex);
```
[the remark this answers is not in the document]
Comment (denis-tikhomirov) [reply src/Agent.Listener/Agent.cs]: Done. Moved this to handler.

File: src/Agent.Listener/Agent.cs [hunk 3]
```diff
@@ -464,6 +464,26 @@ private async Task<int> RunAsync(AgentSettings settings, bool runOnce = false)
...
+                            if (e is AggregateException)
+                            {
+                                Trace.Error("One or several exceptions have been occurred.");
+
+                                int i = 0;
+                                foreach (var ex in ((AggregateException)e).Flatten().InnerExceptions)
+                                {
+
+                                    i++;
+                                    Trace.Error($"InnerException #{i}");
```
[the remark this answers is not in the document]
Comment (denis-tikhomirov) [reply src/Agent.Listener/Agent.cs]: Removed indexing

**Comment by denis-tikhomirov** [comment]
Using of IF-ELSE statements inside of CATCH were replaced with using of several CATCH to handle different types of exceptions.

File: src/Agent.Sdk/Util/ExceptionsUtil.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.VisualStudio.Services.Agent;
+
+namespace Agent.Sdk.Util
+{
+    public class ExceptionsUtil
+    {
+        public static void HandleAggregateException(AggregateException e, Action<string> action)
```
Comment (anatolybolshakov) [inline src/Agent.Sdk/Util/ExceptionsUtil.cs:13]: Nit: please specify more exact name for action delegate - probably something like traceErrorAction?

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks! Let's make buddy testing as well for this one.

File: src/Agent.Sdk/Util/ExceptionsUtil.cs (hunk 4 above)
Comment (denis-tikhomirov) [reply src/Agent.Sdk/Util/ExceptionsUtil.cs]: Done. Renamed it.

**Review by EzzhevNikita — APPROVED** [review]

**Pull request merged**